WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/addons/order/OrderedAddonViewModel.kt

```diff
@@ -59,46 +58,64 @@ class OrderedAddonViewModel @Inject constructor(
     }
 
     private fun mapAddonsFromOrderAttributes(
-        productAddons: List<ProductAddon>,
+        productAddons: List<Addon>,
         orderAttributes: List<Order.Item.Attribute>
-    ) = orderAttributes.mapNotNull { it.findMatchingAddon(productAddons) }
+    ): List<Addon> = orderAttributes.mapNotNull { findMatchingAddon(it, productAddons) }
 
-    private fun Order.Item.Attribute.findMatchingAddon(
-        addons: List<ProductAddon>
-    ) = addons.find { it.name == addonName }
-        ?.asAddonWithSingleSelectedOption(this)
+    private fun findMatchingAddon(matchingTo: Order.Item.Attribute, addons: List<Addon>): Addon? = addons.firstOrNull {
+        it.name == matchingTo.addonName
+    }?.asAddonWithSingleSelectedOption(matchingTo)
 
-    private fun ProductAddon.asAddonWithSingleSelectedOption(
+    private fun Addon.asAddonWithSingleSelectedOption(
         attribute: Order.Item.Attribute
-    ) = options.find { it.label == attribute.value }
-        ?.let { this.copy(rawOptions = listOf(it)) }
-        ?: mergeOrderAttributeWithAddon(this, attribute)
+    ): Addon {
+        return if (this is Addon.HasOptions) {
+            options.find { it.label == attribute.value }
+                ?.let {
+                    when (this) {
+                        is Addon.Checkbox -> this.copy(options = listOf(it))
+                        is Addon.MultipleChoice -> this.copy(options = listOf(it))
+                        else -> null
+                    }
+                } ?: mergeOrderAttributeWithAddon(this, attribute)
+        } else {
+            this
+        }
+    }
```

Comment:
I think some method extractions here would make it easier to understand the intended business logic, the operation looks complex as it is, maybe something like this:

```suggestion
            private fun Addon.asAddonWithSingleSelectedOption(
        attribute: Order.Item.Attribute
    ): Addon {
        return when (this) {
            is Addon.HasOptions -> options
                .find { it.label == attribute.value }
                ?.takeIf { (this is Addon.MultipleChoice) or (this is Addon.Checkbox) }
                ?.let { this.asSelectableAddon(it) }
                ?: mergeOrderAttributeWithAddon(this, attribute)
            else -> this
        }
    }
    
    private fun Addon.asSelectableAddon(selectedOption: Option): Addon? =
        when (this) {
            is Addon.Checkbox -> this.copy(options = listOf(selectedOption))
            is Addon.MultipleChoice -> this.copy(options = listOf(selectedOption))
            else -> null
        }
```

WDYT?

Reply:
Awesome, that's much more readable. Thank you, it was basically copy-paste 🙇 . Added in https://github.com/woocommerce/woocommerce-android/pull/4751/commits/089a5812208672ba0f4ce19378045c21c4864e49